chainntnfs/txconfnotifier.go

```diff
@@ -0,0 +1,239 @@
+		return fmt.Errorf("Received blocks out of order: "+
+			"current height=%d, new height=%d",
+			tcn.currentHeight, blockHeight)
+	}
+	tcn.currentHeight++
+	tcn.reorgDepth = 0
+
+	// Record any newly confirmed transactions in ntfnsByConfirmHeight so that
+	// notifications get dispatched when the tx gets sufficient confirmations.
+	// Also record txs in confTxsByInitialHeight so reorgs can be handled
+	// correctly.
+	for _, tx := range txns {
+		txHash := tx.Hash()
+		for _, ntfn := range tcn.confNotifications[*txHash] {
+			ntfn.details = &TxConfirmation{
+				BlockHash:   blockHash,
+				BlockHeight: blockHeight,
+				TxIndex:     uint32(tx.Index()),
+			}
+
+			confHeight := blockHeight + ntfn.NumConfirmations - 1
+			tcn.ntfnsByConfirmHeight[confHeight] =
+				append(tcn.ntfnsByConfirmHeight[confHeight], ntfn)
+			tcn.confTxsByInitialHeight[blockHeight] =
+				append(tcn.confTxsByInitialHeight[blockHeight], tx.Hash())
+		}
+	}
+
+	// Dispatch notifications for all transactions that are considered confirmed
+	// at this new block height.
+	for _, ntfn := range tcn.ntfnsByConfirmHeight[tcn.currentHeight] {
+		ntfn.Event.Confirmed <- ntfn.details
+		ntfn.dispatched = true
+	}
+
+	// Clear entries from confNotifications and confTxsByInitialHeight. We
+	// assume that reorgs deeper than the reorg safety limit do not happen, so
+	// we can clear out entries for the block that is now mature.
+	if tcn.currentHeight >= tcn.reorgSafetyLimit {
+		matureBlockHeight := tcn.currentHeight - tcn.reorgSafetyLimit
+		for _, txHash := range tcn.confTxsByInitialHeight[matureBlockHeight] {
+			delete(tcn.confNotifications, *txHash)
+		}
+		delete(tcn.confTxsByInitialHeight, matureBlockHeight)
+	}
+
+	return nil
+}
+
+// DisconnectTip handles the tip of the current chain being disconnected during
+// a chain reorganization. If any watched transactions were included in this
+// block, internal structures are updated to ensure a confirmation notification
+// is not sent unless the transaction is included in the new chain.
+func (tcn *TxConfNotifier) DisconnectTip(blockHeight uint32) error {
+	if blockHeight != tcn.currentHeight {
+		return fmt.Errorf("Received blocks out of order: "+
+			"current height=%d, disconnected height=%d",
+			tcn.currentHeight, blockHeight)
+	}
+	tcn.currentHeight--
+	tcn.reorgDepth++
+
+	for _, txHash := range tcn.confTxsByInitialHeight[blockHeight] {
+		for _, ntfn := range tcn.confNotifications[*txHash] {
+			// If notification has been dispatched with sufficient
+			// confirmations, notify of the reversal.
+			if ntfn.dispatched {
+				ntfn.Event.NegativeConf <- int32(tcn.reorgDepth)
+				ntfn.dispatched = false
+				continue
+			}
+
+			confHeight := blockHeight + ntfn.NumConfirmations - 1
+
+			// Remove ntfn from ntfnsByConfirmHeight.
+			idx := -1
+			slice := tcn.ntfnsByConfirmHeight[confHeight]
+			for i, entry := range slice {
+				if entry == ntfn {
+					idx = i
+					break
+				}
+			}
+
+			if idx == -1 {
```

Comment:
This if else statement needs some more comments, not obvious what's going on imho :P